File: tests/workflow/producer/rust/src/producer/responses/groupa_structb.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use crate::stat::Alias;
use crate::test::samples;
use clibri::server;
use uuid::Uuid;

type BroadcastGroupBGroupCStructB = (Vec<Uuid>, protocol::GroupB::GroupC::StructB);

pub enum Response {
    GroupBStructA((protocol::GroupB::StructA, BroadcastGroupBGroupCStructB)),
    GroupBGroupCStructA(protocol::GroupB::GroupC::StructA),
}

#[allow(unused_variables)]
pub async fn response<'c, E: server::Error, C: server::Control<E>>(
    request: &protocol::GroupA::StructB,
    scope: &mut Scope<'_, E, C>,
) -> Result<Response, protocol::GroupA::StructB> {
    let index = scope
        .context
        .get_index(scope.identification.uuid(), Alias::GroupAStructB);
    if index == 1 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::GroupBStructA);
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::GroupBGroupCStructB);
        Ok(Response::GroupBStructA((
            samples::group_b::struct_a::get(),
            (
                vec![scope.identification.uuid()],
                samples::group_b::group_c::struct_b::get(),
            ),
        )))
    } else if index == 2 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::GroupBGroupCStructA);
        Ok(Response::GroupBGroupCStructA(
            samples::group_b::group_c::struct_a::get(),
        ))
    } else {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::GroupAStructB);
        Err(samples::group_a::struct_b::get())
    }
}

File: tests/workflow/producer/typescript/src/producer/responses/groupa.structb.ts
import { Response } from "../implementation/responses/groupa.structb";
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/responses";
import { Alias } from "../../stat";
import { panic } from "../../tools";
import { Scope } from "../implementation/scope";

export function response(
	request: Protocol.GroupA.StructB,
	scope: Scope
): Promise<Response> {
	const stat = scope.context.getStat(scope.consumer.uuid());
	stat.case(Alias.GroupAStructB);
	const index = stat.index(Alias.GroupAStructB);
	if (index === 0) {
		return Promise.resolve(
			new Response(Protocol.GroupB.StructA.defaults())
				.broadcast([scope.consumer.uuid()])
				.GroupBGroupCStructB(Protocol.GroupB.GroupC.StructB.defaults())
		);
	} else if (index === 1) {
		return Promise.resolve(
			new Response(Protocol.GroupB.GroupC.StructA.defaults())
		);
	} else if (index === 2) {
		return Promise.resolve(
			new Response(Protocol.GroupA.StructB.defaults())
		);
	} else {
		panic(`Unexpected index for Protocol.GroupA.StructB request: ${index}`);
		return Promise.reject(
			new Error(`Handler for Protocol.GroupA.StructB isn't implemented.`)
		);
	}
}
